{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "## Lab_6. Differential RNA expression analysis\n",
        "\n",
        "In this project we will study changes that happen in yeast cells before or during fermentation and see in-detail the changes in RNA expression during the fermentation process.\n",
        "\n",
        "\n",
        "[Link on results files ](https://drive.google.com/drive/folders/1lloDZ8gyPw7AWWTsxtzn7eDc9H3VVZW7)"
      ],
      "metadata": {
        "id": "cFkOrL34obBz"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### 1. Input data\n",
        "\n",
        "There are two replicates of RNA-seq data from yeast before and during\n",
        "fermentation, and my goal is to find out if the yeast express different genes during fermentation than they do under normal growth.\n",
        "\n",
        "\n",
        "**yeast reads:**\n",
        "\n",
        "**SRR941816: fermentation 0 minutes replicate 1**\n",
        "ftp.sra.ebi.ac.uk/vol1/fastq/SRR941/SRR941816/SRR941816.fastq.gz (413 Mb)\n",
        "\n",
        "**SRR941817: fermentation 0 minutes replicate 2**\n",
        "ftp.sra.ebi.ac.uk/vol1/fastq/SRR941/SRR941817/SRR941817.fastq.gz (455 Mb)\n",
        "\n",
        "**SRR941818: fermentation 30 minutes replicate 1**ftp.sra.ebi.ac.uk/vol1/fastq/SRR941/SRR941818/SRR941818.fastq.gz (79.3 Mb)\n",
        "\n",
        "**SRR941819: fermentation 30 minutes replicate 2**\n",
        "ftp.sra.ebi.ac.uk/vol1/fastq/SRR941/SRR941819/SRR941819.fastq.gz\n",
        "\n",
        "---\n",
        "\n",
        "\n",
        "\n",
        "As a reference genome we will use Saccharomyces cerevisiae, in the genome database at NCBI., strain S288c and assembly R64. Download the reference genome in FASTA format and annotation in GFF format.\n",
        "\n",
        "**reference genome file**\n",
        "ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.fna.gz\n",
        "\n",
        "**annotation file**\n",
        "ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_genomic.gff.gz"
      ],
      "metadata": {
        "id": "gRrifarjtfnK"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### 2. Analysis Pipeline\n",
        "\n",
        "**Analysis flow chart**\n",
        "\n",
        "**Inputs** - *fastq + reference genome + annotation*\n",
        "\n",
        "⇓\n",
        "\n",
        "**Pre-processing (indexing)**- *HISAT2*\n",
        "\n",
        "⇓\n",
        "\n",
        "**Alignement** - *HISAT2*\n",
        "\n",
        "⇓\n",
        "\n",
        "**Abundance estimation** - *featureCounts*\n",
        "\n",
        "⇓\n",
        "\n",
        "**Expression analysis and Visualization** - *DESeq2, gplots*\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "bgRcactsu-jj"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "####**a) Aligning with HISAT2**\n",
        "\n",
        "build genome index:\n",
        "\n",
        "```bash\n",
        "gunzip GCF_000146045.2_R64_genomic.fna.gz\n",
        "ln -s GCF_000146045.2_R64_genomic.fna reference\n",
        "hisat2-build -p 4 reference genome_index\n",
        "```\n",
        "run hisat2 in single-end mode:\n",
        "\n",
        "``` python\n",
        "SAMPLES = [\"SRR941816\", \"SRR941817\", \"SRR941818\", \"SRR941819\"]\n",
        "HISAT2_INDEX = \"/home/anna/BioInf_practice/6/genome_index\"\n",
        "THREADS = 4\n",
        "\n",
        "rule all:\n",
        "    input:\n",
        "        expand(\"aligned/{sample}.bam\", sample=SAMPLES)\n",
        "\n",
        "rule extract_fastq:\n",
        "    input:\n",
        "        \"{sample}.fastq.gz\"\n",
        "    output:\n",
        "        \"unpacked/{sample}.fastq\"\n",
        "    shell:\n",
        "        \"gunzip -c {input} > {output}\"\n",
        "\n",
        "rule align_and_sort:\n",
        "    input:\n",
        "        unpacked_fastq=\"unpacked/{sample}.fastq\"\n",
        "    output:\n",
        "        bam=\"aligned/{sample}.bam\"\n",
        "    params:\n",
        "        index=HISAT2_INDEX\n",
        "    threads: THREADS\n",
        "    shell:\n",
        "        \"\"\"\n",
        "        hisat2 -p {threads} -x {params.index} -U {input.unpacked_fastq} | \\\n",
        "        samtools sort -@ {threads} -o {output.bam}\n",
        "        \"\"\"\n",
        "snakemake --cores 4\n",
        "```\n",
        "####**b) Quantifying with featureCounts**\n",
        "\n",
        "featureCounts can not work with GFF files. I converted the GFF file to GTF format with gffread.\n",
        "\n",
        "```bash\n",
        "mamba install gffread\n",
        "ln -s GCF_000146045.2_R64_genomic.gff reference_gff\n",
        "gffread GCF_000146045.2_R64_genomic.gff -T -o GCF_000146045.2_R64_genomic.gtf\n",
        "```\n",
        "\n",
        "To run the feature counts program I add 2 new rules to snakemake pipline:\n",
        "\n",
        "```python\n",
        "ANNOTATION_FILE = \"/home/anna/BioInf_practice/6/reference_gtf\"\n",
        "rule all:\n",
        "    input:\n",
        "        expand(\"simple_counts/{sample}_simple_counts.txt\", sample=SAMPLES)\n",
        "        \n",
        "rule featureCounts:\n",
        "    input:\n",
        "        bam=\"aligned/{sample}.bam\"\n",
        "    output:\n",
        "        counts=\"counts/{sample}_counts.txt\"\n",
        "    params:\n",
        "        annotation=ANNOTATION_FILE\n",
        "    shell:\n",
        "        \"\"\"\n",
        "        featureCounts -g gene_id -a {params.annotation} -o {output.counts} {input.bam}\n",
        "        \"\"\"\n",
        "\n",
        "rule simplify_counts:\n",
        "    input:\n",
        "        counts=\"counts/{sample}_counts.txt\"\n",
        "    output:\n",
        "        simple_counts=\"simple_counts/{sample}_simple_counts.txt\"\n",
        "    shell:\n",
        "        \"\"\"\n",
        "        tail -n +2 {input.counts} | cut -f 1,7- > {output.simple_counts}\n",
        "        \"\"\"\n",
        "```\n",
        "And the pipline above also include the choice of necessary columns from the output files.\n"
      ],
      "metadata": {
        "id": "nUUhpn2gx3KE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "####**c) Find differentially expressed genes with Deseq2**\n",
        "\n",
        "To calculate metrics and build a heatmap I used the following ran the following R scripts in R studio\n",
        "\n",
        "```r\n",
        "library(ggplot2)\n",
        "library(pheatmap)\n",
        "library(DESeq2)\n",
        "\n",
        "\n",
        "SRR16 <- read.table(file(\"SRR941816_simple_counts.txt\"), header=TRUE, row.names=1, sep=\"\\t\")\n",
        "SRR17 <- read.table(file(\"SRR941817_simple_counts.txt\"), header=TRUE, row.names=1, sep=\"\\t\")\n",
        "SRR18 <- read.table(file(\"SRR941818_simple_counts.txt\"), header=TRUE, row.names=1, sep=\"\\t\")\n",
        "SRR19 <- read.table(file(\"SRR941819_simple_counts.txt\"), header=TRUE, row.names=1, sep=\"\\t\")\n",
        "\n",
        "combinedCountData <- cbind(SRR16, SRR17, SRR18, SRR19)\n",
        "colnames(combinedCountData) <- c(\"SRR941816_control\", \"SRR941817_control\", \"SRR941818_exposure\", \"SRR941819_exposure\")\n",
        "conditions <- c(\"control\", \"control\", \"ferm30\", \"ferm30\")\n",
        "sample_info <- data.frame(Condition = conditions)\n",
        "rownames(sample_info) <- c(\"SRR941816_control\", \"SRR941817_control\", \"SRR941818_exposure\", \"SRR941819_exposure\")\n",
        "\n",
        "sample_info$Condition <- factor(sample_info$Condition)\n",
        "\n",
        "# DESeqDataSet object creation\n",
        "dds <- DESeqDataSetFromMatrix(countData = combinedCountData,\n",
        "                              colData = sample_info,\n",
        "                              design = ~Condition)\n",
        "dds$Condition <- factor(dds$Condition, levels = c('control', 'ferm30'))\n",
        "dds <- DESeq(dds)\n",
        "\n",
        "# Obtaining the results\n",
        "res <- results(dds)\n",
        "\n",
        "# Step 1. Results filtering\n",
        "filtered_genes <- rownames(res)[res$baseMean > 0  & complete.cases(res) & res$padj < 0.05]\n",
        "\n",
        "# Step 2: dds filtering\n",
        "dds_filtered <- dds[filtered_genes,]\n",
        "res_filtered <-  results(dds_filtered)\n",
        "normalized_counts <- counts(dds_filtered, normalized=TRUE)\n",
        "\n",
        "# Step 3: heatmap building\n",
        "pheatmap(log1p(normalized_counts),\n",
        "         clustering_distance_rows = \"euclidean\",\n",
        "         clustering_distance_cols = \"euclidean\",  \n",
        "         clustering_method = \"complete\",  \n",
        "         show_rownames = FALSE,\n",
        "         scale = 'row')\n",
        "```\n",
        "\n"
      ],
      "metadata": {
        "id": "ZnapW_EtbvNN"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#### **3. Result Interpretation**\n",
        "\n",
        "Then I took top-50 genes from the results sorted by sorted by adjusted p-values and saved them to a txt file to map them through GO (gene ontology) database.\n",
        "\n",
        "\n",
        "```r\n",
        "library(dplyr)\n",
        "res_df <- as.data.frame(res_filtered)\n",
        "res_sorted <- res_df %>%\n",
        "  arrange(padj) %>%\n",
        "  head(50)\n",
        "top50_genes <- rownames(res_sorted)[1:50]\n",
        "writeLines(top50_genes, \"top50_genes.txt\")\n",
        "top50_dds <- dds_filtered[top50_genes, ]\n",
        "res_top50 <- results(top50_dds)\n",
        "normalized_counts_top50 <-counts(top50_dds, normalized=TRUE)\n",
        "pheatmap(log1p(normalized_counts_top50),\n",
        "         clustering_distance_rows = \"euclidean\",\n",
        "         clustering_distance_cols = \"euclidean\",\n",
        "         clustering_method = \"complete\",  \n",
        "         show_rownames = TRUE)\n",
        "\n",
        "```\n",
        "After mapping selected genes to GO database results interpretation was started."
      ],
      "metadata": {
        "id": "cMetXNhzqIkY"
      }
    }
  ]
}